include/garden/system/timer.hpp

```diff
@@ -0,0 +1,145 @@
+/**
+ * \file timer.hpp
+ * \author Sahak Grigoryan <sahakgrigoryan.am@gmail.com>
+ * 
+ * \brief Class definition: TimerSystem
+ */
+
+#pragma once
+
+//--------------------------------------------------------------------------
+// includes
+
+// C/C++ standard headers
+#include <chrono>
+#include <unordered_set>
+
+// third-party library headers
+#include "ecsm.hpp"
+
+// project-specific headers
+
+//--------------------------------------------------------------------------
+// forward declarations
+namespace ecsm
+{
+template <class, bool>
+class LinearPool;
+} // namespace ecsm
+
+//--------------------------------------------------------------------------
+// class definition and member declarations
+namespace garden
+{
+
+using ecsm::IdHash;
+using ecsm::ID;
+using ecsm::Singleton;
+using ecsm::System;
+using ecsm::View;
+
+/**
+ * \class TimerSystem
+ * \brief Manages timer-based event handlers.
+ */
+class TimerSystem final : public System, public Singleton<TimerSystem>
+{
+public:
+    using ClockType     = std::chrono::steady_clock;
+    using TimeUnit      = std::chrono::milliseconds;
+    using TimePoint     = std::chrono::time_point<ClockType, TimeUnit>;
+    using TimerCallback = std::function<bool(void)>;
+
+    /**
+     * \brief Timer metadata structure.
+     */
+    struct TimerHandler {
```

Comment:
In Entity Component System (ECS) paradigm this should be a TimerComponent, please see the ComponentSystem<> usage. With current implementation we can't attach this timer to an entity.